src/sif/autocost.cc

```diff
@@ -425,7 +425,18 @@ bool AutoCost::Allowed(const baldr::DirectedEdge* edge,
       edge->surface() == Surface::kImpassable || IsUserAvoidEdge(edgeid) ||
       (!allow_destination_only_ && !pred.destonly() && edge->destonly()) ||
       (pred.closure_pruning() && IsClosed(edge, tile)) ||
-      (exclude_unpaved_ && !pred.unpaved() && edge->unpaved()) || !IsHOVAllowed(edge)) {
+      (exclude_unpaved_ && !pred.unpaved() && edge->unpaved()) || !IsHOVAllowed(edge) ||
```

Comment:
can we please:
- initialize a `has_excludes` or so at the beginning of each cost model which is true if _any_ of them is set? this is just way too many checks right now for _every_ edge which are basically never triggered on average
- make this a function in DynamicCost, it's used in every cost model, avoids a lot of duplication

Reply:
what about shortcuts btw?

Reply:
@nilsnolde for shortcuts we do check if edge is shortcut and exclude_bridge/tunnel is true then we do not use it. As per your discussion in [#3733](https://github.com/valhalla/valhalla/pull/3733) you decided on this change! Can you please specify exactly what could be the issue with shortcuts if we exclude edges part of shortcut?

Reply:
https://github.com/valhalla/valhalla/pull/3733#issuecomment-1235883038 here its mentioned that we can disable shortcut use

Reply:
the initial PR is 1.5 years old, I have a hard time remembering what I did last week;) that's the problem with opening new PRs based on old ones, I won't go back and read across several PRs. that's why it's important for the person who raises a PR to be as informative as needed for reviewers to have an easy-ish time. 

now that you pointed it out I squinted again at the code and yeah I see the last line in each cost model update handles shortcuts in a way that should be fine (I think).

that's kinda a case in point of above though: those changes are so repetitive it's very easy to miss these small important updates in a review and it'd be very nice to have it commented in the PR so those misunderstandings don't happen. keep in mind we/I do this on a voluntary basis and the easier you make the reviewer's lives the likelier they'll actually put in the unpaid time to do the review. just as a recent example, e.g. in https://github.com/valhalla/valhalla/pull/4516 one can see how reviewers will have an easier time when the PR person is flagging non-trivial changes explicitly and explaining a little bit right where the changes happen (not in the PR description).

Reply:
> can we please:
> 
> * initialize a `has_excludes` or so at the beginning of each cost model which is true if _any_ of them is set? this is just way too many checks right now for _every_ edge which are basically never triggered on average

This is Done

> * make this a function in DynamicCost, it's used in every cost model, avoids a lot of duplication

This is Done

Regarding the shortcuts i guess its disabled correctly as mentioned in [Comment](https://github.com/valhalla/valhalla/pull/3733#issuecomment-1242590288)

As you have asked, i commented on the files in the PR for easier review

Reply:
ok yeah i mentioned that one could disable it on costing but that is really really going to be a performance hit. i wonder how bad it would be all the shortcut builder to care about tunnel and bridge attribution. im tempted to think it might not be so bad compared to the alternative here. an experiment would be worth it at some point and if it proves not too bad we could replace the costing skipping shortcuts with that